pkg/controller/common.go

```diff
@@ -1259,3 +1320,64 @@ func containsString(key string, items []string) bool {
 	}
 	return false
 }
+
+type ReplicaRequest struct {
+	volumeID      string
+	volumeContext map[string]string //change this to include actual information
+	priority      int               //The number of replicas that have yet to be created
+	index         int
+}
+
+type VolumeReplicaRequestsPriorityQueue struct {
+	mut   *sync.Mutex
+	queue PriorityQueue
+}
+
+func (vq VolumeReplicaRequestsPriorityQueue) Push(replicaRequest ReplicaRequest) {
+	vq.mut.Lock()
+	defer vq.mut.Unlock()
+	vq.queue.Push(replicaRequest)
+	return
+}
+
+func (vq VolumeReplicaRequestsPriorityQueue) Pop() *ReplicaRequest {
+	vq.mut.Lock()
+	defer vq.mut.Unlock()
+	return vq.queue.Pop().(*ReplicaRequest)
+}
+
+type PriorityQueue []*ReplicaRequest
```

Comment:
It makes sense to create a generic priority queue implementation in `common.go` or even `pkg/util` rather that tie it specifically to `ReplicaRequest`. You can also omit the self-index tracking and update mechanism that the example has since we do not need it.

```go
type LessFunc func(left, right interface{}) bool

type itemsArray struct {
	lessFn LessFunc
	items  []interface{}
}

func (i *itemsArray) Len() int {
	return len(i.items)
}

func (i *itemsArray) Less(left, right int) bool {
	return i.lessFn(i.items[left], i.items[right])
}

func (i *itemsArray) Swap(x, y int) {
	i.items[x], i.items[y] = i.items[y], i.items[x]
}

func (i *itemsArray) Push(value interface{}) {
	i.items = append(i.items, value)
}

func (i *itemsArray) Pop() interface{} {
	old := i.items
	n := len(old)
	value := old[n-1]
	old[n-1] = nil
	i.items = old[0 : n-1]

	return value
}

type PriorityQueue struct {
	items itemsArray
}

func NewPriorityQueue(lessFn LessFunc) *PriorityQueue {
	pq := &PriorityQueue{items: itemsArray{lessFn: lessFn}}

	heap.Init(&pq.items)

	return pq
}

func (pq *PriorityQueue) Push(value interface{}) {
	heap.Push(&pq.items, value)
}

func (pq *PriorityQueue) Pop() interface{} {
	return heap.Pop(&pq.items)
}
```

With this implementation, you create a priority queue for `ReplicaRequest` like this:

```go
queue := NewPriorityQueue(func (left, right interface{}) bool {
  return left.(*ReplicaRequest).priority > right.(*ReplicaRequest).priority
})
```